WebGPU API docs (#24202)
mdn/content: files/en-us/web/api/webgpu_api/index.md

> ---

Reviewer: I didn't re-review this file, but if there's updates to it that you want me to look at let me know!

Author: Nothing significant that I think you need to see; just a couple of small updates to fill in descriptions of interfaces

Author: I have updated the interface descriptions, both in light of comments made in this review, and also to provide descriptions that were previously just placeholders.

I have also grouped the interfaces into different categories. We don't normally do this, but in this case the API is so big that I thought it would be impossible to follow if we just has one big alphabetical list.

I'd love you to check the "Interfaces" section out and let me know what you think.

Reviewer: On it!

However, WebGL still has some fundamental issues that need addressing:

- Since WebGL's release, a new generation of native GPU APIs have appeared — the most popular being [Microsoft's Direct3D 12](https://docs.microsoft.com/en-us/windows/win32/direct3d12/direct3d-12-graphics), [Apple's Metal](https://developer.apple.com/metal/), and [The Khronos Group's Vulkan](https://www.vulkan.org/) — which provide a multitude of new features. There are no more updates planned to OpenGL (and therefore WebGL), therefore it won't get any of these new features. WebGPU on the other hand will have new features added to it going forwards.
- WebGL is based wholly around the use case of drawing graphics and rendering them to a canvas. It does not handle general-purpose GPU (GPGPU) computations very well. GPGPU computations are becoming more and more important for many different use cases, for example those based on machine learning models.
- 3D graphics apps are becoming increasingly demanding, both in terms of the number of objects to be rendered simultaneously, and usage of new rendering features.

> WebGPU addresses these issues, providing an updated general-purpose architecture compatible with modern GPU APIs, which feels more "webby". It supports graphic rendering, but also has first-class support for GPGPU computations. Rendering of individual objects is significantly cheaper on the CPU side, and it supports modern GPU rendering features such as compute-based particles and post-processing filters like color effects, sharpening, and depth-of-field simulation. In addition, it can handle expensive computations such as culling and skinned model transformation.

Reviewer 2: ```suggestion
WebGPU addresses these issues, providing an updated general-purpose architecture compatible with modern GPU APIs, which feels more "webby". It supports graphic rendering, but also has first-class support for GPGPU computations. Rendering of individual objects is significantly cheaper on the CPU side, and it supports modern GPU rendering features such as compute-based particles and post-processing filters like color effects, sharpening, and depth-of-field simulation. In addition, it can handle expensive computations directly on the GPU such as culling and skinned model transformation.
```

Reviewer 2: WebGL could already do these on the CPU, what's new is that WebGPU makes it tractable to defer these to the GPU.

Author: Added to my next commit

   4. Finalize the command list and encapsulate it in a command buffer.
   5. Submit the command buffer to the GPU via the logical device's command queue.

In the sections below, we will examine a basic render pipeline demo, to allow you to explore what a pipeline requires. Later on, we'll also examine a [basic compute pipeline](#basic_compute_pipeline) example, looking at how it differs from the render pipeline.

## Basic render pipeline

In our [basic render demo](https://webgpu-basic-render.glitch.me/) we give a `<canvas>` element a solid blue background and draw a triangle onto it.

### Create shader modules

We are using the following shader code. The vertex shader stage (`@vertex` block) accepts a chunk of data containing a position and a color, positions the vertex according to the given position, interpolates the color, then passes the data along to the fragment shader stage. The fragment shader stage (`@fragment` block) accepts the data from the vertex shader stage and colors the vertex according to the given color.

```js

> const shaders = `
> struct VertexOut {
>     @builtin(position) position : vec4f,
>     @location(0) color : vec4f
> }
> 
> @vertex
> fn vertex_main(@location(0) position: vec4f,
>                  @location(1) color: vec4f) -> VertexOut
> {

Reviewer 2: This is making indentation consistent.
```suggestion
const shaders = `
struct VertexOut {
  @builtin(position) position : vec4f,
  @location(0) color : vec4f
}

@vertex
fn vertex_main(@location(0) position: vec4f,
               @location(1) color: vec4f) -> VertexOut
{
```

Author: Ah, good call. I've gone through all the code and made sure the indentation is consistent, in the article and demos.

The {{domxref("GPUBuffer")}} is created via a call to {{domxref("GPUDevice.createBuffer()")}}. We give it a size equal to the length of the `vertices` array so it can contain all the data, `VERTEX` and `COPY_DST` flags to indicate that the buffer will be used as a vertex buffer and the destination of copy operations.

```js
const vertexBuffer = device.createBuffer({
    size: vertices.byteLength, // make it big enough to store vertices in
    usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
});
```

We could handle getting our data into the `GPUBuffer` using a mapping operation, like we use in the [compute pipeline example](#basic_compute_pipeline) to read data from the GPU back to JavaScript. However, in this case we are going to use the handy {{domxref("GPUQueue.writeBuffer()")}} convenience method, which takes as its parameters the buffer to write to, the data source to write from, and offet value for each, and the size of data to write (we've specified the whole length of the array). The browser then works out the most efficient way to handle writing the data.

```js

> const array = new Float32Array(vertices);

Reviewer: Vertices are already in a Float32Array, so you can remove this line and use `vertices` below. (In fact, you've got a small bug currently in that you're using `vertices.length` but passing `array`. They work out to the same thing here, but in a more complex scenario that would likely break.)

Author: Ah, good catch. Fixed!

These three steps can be achieved via the following two lines:

```js
passEncoder.end();

device.queue.submit([commandEncoder.finish()]);
```

## Basic compute pipeline

In our [basic compute demo](https://webgpu-basic-compute.glitch.me/), we get the GPU to calculate some values, store them in an output buffer, copy the data across to a staging buffer, then map that staging buffer so that the data can be read out to JavaScript and logged to the console.

The app follows a similar structure to the basic rendering demo. We create a {{domxref("GPUDevice")}} reference in the same way as before, and encapsulate our shader code into a {{domxref("GPUShaderModule")}} via a {{domxref("GPUDevice.createShaderModule()")}} call. The difference here is that our shader code only has one shader stage, a compute shader:

```js
const shader = `
@group(0) @binding(0)
var<storage, read_write> output: array<f32>;

@compute @workgroup_size(64)
fn main(

  @builtin(global_invocation_id)
  global_id : vec3u,

  @builtin(local_invocation_id)
  local_id : vec3u,

) {

>   output[global_id.x] =
>     f32(global_id.x) * 1000. + f32(local_id.x);

Reviewer 2: ```suggestion
  // Avoid accessing the buffer out of bounds.
  if (global_id.x >= ${BUFFER_SIZE}) {
    return;
  }
  
  output[global_id.x] =
    f32(global_id.x) * 1000. + f32(local_id.x);
```

Author: Ah, this makes sense. I've added it to the article and compute demo. Thanks!

```js
const bindGroup = device.createBindGroup({
    layout: bindGroupLayout,
    entries: [{
        binding: 0,
        resource: {
            buffer: output,
        }
    }]
});
```

> **Note:** You could retrieve an implicit layout to use when creating a bind group by calling the {{domxref("GPUComputePipeline.getBindGroupLayout()")}} method. There is also a version available for render pipelines: see {{domxref("GPURenderPipeline.getBindGroupLayout()")}}.

### Create a compute pipeline

With the above all in place, we can now create a compute pipeline by calling {{domxref("GPUDevice.createComputePipeline()")}}, passing it a pipeline descriptor object. This works in a similar way to creating a render pipeline. We describe the compute shader, specifying what module to find the code in and what the entry point is. We also specify a `layout` for the pipeline, in this case creating a layout based on the `bindGroupLayout` we defined earlier via a {{domxref("GPUDevice.createPipelineLayout()")}} call.

```js
const computePipeline = device.createComputePipeline({
    layout: device.createPipelineLayout({

>         bindGroupLayouts: [bindGroupLayout]

Reviewer 2: ```suggestion
      bindGroupLayouts: [bindGroupLayout]
```

## Concepts and usage

It is fair to say that {{domxref("WebGL_API", "WebGL", "", "nocode")}} revolutionized the web in terms of graphical capabilities after it first appeared around 2011. WebGL is a JavaScript port of the [OpenGL ES 2.0](https://registry.khronos.org/OpenGL-Refpages/es2.0/) graphics library, allowing web pages to pass rendering computations directly to the device's GPU to be processed at very high speeds, and render the result inside a {{htmlelement("canvas")}} element.

WebGL and the [GLSL](https://www.khronos.org/opengl/wiki/Core_Language_(GLSL)) language used to write WebGL shader code are complex, so several WebGL libraries were created to make WebGL apps easier to write. Popular examples include [Three.js](https://threejs.org/), [Babylon.js](https://www.babylonjs.com/), and [PlayCanvas](https://playcanvas.com/). Developers have used these tools to build impressive web-based 3D games, music videos, training and modeling tools, VR and AR experiences, and more.

However, WebGL still has some fundamental issues that need addressing:

> - Since WebGL's release, a new generation of native GPU APIs have appeared — the most common being [Microsoft's Direct3D 12](https://docs.microsoft.com/en-us/windows/win32/direct3d12/direct3d-12-graphics), [Apple's Metal](https://developer.apple.com/metal/), and [The Khronos Group's Vulkan](https://www.vulkan.org/) — which provide a multitude of new features. WebGL is not able to access many of these due to being based on an old-fashioned Open GL architecture.

Author: I've made some updates to address this.

> - Since WebGL's release, a new generation of native GPU APIs have appeared — the most common being [Microsoft's Direct3D 12](https://docs.microsoft.com/en-us/windows/win32/direct3d12/direct3d-12-graphics), [Apple's Metal](https://developer.apple.com/metal/), and [The Khronos Group's Vulkan](https://www.vulkan.org/) — which provide a multitude of new features. WebGL is not able to access many of these due to being based on an old-fashioned Open GL architecture.

Reviewer 2: Technically WebGL can still be updated a bunch to match current levels of feature of OpenGL but that phrasing is good. Thanks!

- WebGL is based wholly around the use case of drawing graphics and rendering them to a canvas. It does not handle general-purpose GPU (GPGPU) computations very well. GPGPU computations are becoming more and more important for many different use cases, for example those based on machine learning models.
- 3D graphics apps are becoming increasingly demanding, both in terms of the number of objects to be rendered simultaneously, and usage of new rendering features.

> WebGPU addresses these issues, providing a general-purpose architecture compatible with modern GPU APIs, which feels more "webby". It supports graphic rendering, but also has first support for GPGPU computations. Rendering of individual objects is significantly cheaper and it supports modern GPU rendering features such as compute-based particles and post-processing filters like color effects, sharpening, and depth-of-field simulation. In addition, it can handle expensive computations such as culling and skinned model transformation that previously were often done on the CPU.

Reviewer 2: LGTM, with a slight nit on the last sentence.

## General model

There are several layers of abstraction between the device GPU and the web browser running the WebGPU API. It is useful to understand these are you begin to learn WebGPU:

![A basic stack diagram showing the position of the different elements of a WebGPU architecture on a device](basic-webgpu-stack.svg)

> - A physical device has GPUs. Most devices only have one GPU, but some have more than one. GPUs are broken down into two types: Integrated GPUs (which live on the same board as the CPU and share its memory) and discrete GPUs (which live on their own board, separate from the CPU).

Author: I've updated the text to mention this and be a bit more future-proof (e.g. mention that different types are available, not "two types")

> - A physical device has GPUs. Most devices only have one GPU, but some have more than one. GPUs are broken down into two types: Integrated GPUs (which live on the same board as the CPU and share its memory) and discrete GPUs (which live on their own board, separate from the CPU).

Reviewer: LGTM

> - A physical device has GPUs. Most devices only have one GPU, but some have more than one. GPUs are broken down into two types: Integrated GPUs (which live on the same board as the CPU and share its memory) and discrete GPUs (which live on their own board, separate from the CPU).

Reviewer: LGTM

- A software driver allows a GPU to talk to the device OS — it is responsible for sending instructions to the GPU and receiving responses.
- The native OS GPU API, which is part of the OS (e.g. Metal on macOS) is a programming interface allowing native applications to use the capabilities of the GPU. API instructions are sent to the GPU (and responses received) via the driver.

> - An adapter translates instructions between the native OS GPU API and WebGPU.

Author: I think @Kangz made a comment about this too. I'll look at that and make an update in my next commit, and you can let me know what you think.

> - An adapter translates instructions between the native OS GPU API and WebGPU.

Author: I've read through your comments carefully, and I think I get what the issue is here. I've tried rewriting the text and providing a new diagram in my next commit; please let me know what you think. If I'm still missing the mark, perhaps we could have a brief chat to try to hash it out.

> - An adapter translates instructions between the native OS GPU API and WebGPU.

Reviewer 2: I think this is extremely clear now!

> - An adapter translates instructions between the native OS GPU API and WebGPU.

Reviewer: Updated text feels better, thanks!

2. You access an adapter via the {{domxref("GPU.requestAdapter", "GPU.requestAdapter()")}} method. This method accepts an optional settings object allowing you to request for example a high-performance or low-energy adapter. If this is not included, the device will provide access to the default adapter, which is good enough for most purposes.
3. A device can be requested via {{domxref("GPUAdapter.requestDevice()")}}. This method also accepts an options object (referred to as a descriptor), which can be used to specify the exact features and limits you want the logical device to have. If this is not included, the supplied device will have a reasonable general-purpose spec that is good enough for most purposes.

Putting this together with some feature detection checks, the above process could be achieved as follows:

```js
async function init() {
  if (!navigator.gpu) {
    throw Error('WebGPU not supported.');
  }

  const adapter = await navigator.gpu.requestAdapter();
  if (!adapter) {
    throw Error('Couldn\'t request WebGPU adapter.');
  }

>   const device = await adapter.requestDevice();

Author: Ah, yes, @toji also commented on this. I'd appreciate some thoughts on what snippet would be best to replace it at present. I'd really like to include some feature detection. Perhaps the first two blocks would be enough, and I could just omit the `if (!device) { ... }` block?

>   const device = await adapter.requestDevice();

Reviewer 2: I think just omitting the `if (!device)` block would work. TBH I'm not 100% clear on what we expect new WebGPU developers to do: the intent of having failed device creation return a lost device is that it makes it easy for applications to have a single code path that handles failed device creation and actual device loss for unexpected reasons (unplugging and external GPU for example).

>   const device = await adapter.requestDevice();

Author: I think just removing that block is fine; that's what I've done for now. If it transpires that later on, we want to add more info on error handling, we can always do that

  if (!device) {

>     throw Error('Couldn\'t request WebGPU logical device.');

Author: What code would you suggest as a best practice alternative? I've tried playing around a bit to write some code that throws an error if the device is lost, but it is made a bit more complex by the fact that device.lost returns a promise.

  }

  //...

}
```

## Pipelines and shaders: WebGPU app structure

A pipeline is a logical structure containing programmable stages that are completed to get your program's work done. WebGPU is currently able to handle two types of pipeline:

- A render pipeline renders graphics, typically into a {{htmlelement("canvas")}} element, but it could also render graphics offscreen. It has two main stages:

>   - A vertex stage, in which a vertex shader takes positioning data fed into the GPU and uses it to position the vertices of triangles (the basic building block of rendered graphics) in 3D space. The GPU applies other specified effects like rotation, translation, or perspective, then rasterizes the triangles to fingure out what pixels each one should cover on the drawing canvas.

Author: @toji provided a rewrite of this point, which I think fixes the issue you've raised. Let me know if you'd like to see any other changes here.

>   - A fragment stage, in which a fragment shader takes color data and uses it to color in each triangle.

Author: This has also been updated.

>   - A fragment stage, in which a fragment shader takes color data and uses it to color in each triangle.

Reviewer 2: LGTM

> - A compute pipeline is for general computation. A compute pipeline contains a single compute stage, in which a compute shader takes general data, processes it as specified, then returns the result in a buffer. The buffer can contain any kind of data.

Author: Got it. Let me know if you are happy with the updated version.

> - A compute pipeline is for general computation. A compute pipeline contains a single compute stage, in which a compute shader takes general data, processes it as specified, then returns the result in a buffer. The buffer can contain any kind of data.

Reviewer 2: LGTM

The shaders mentioned above are sets of instructions processed by the GPU. WebGPU shaders are written in a low-level Rust-like language called [WebGPU Shader Language](https://gpuweb.github.io/gpuweb/wgsl/) (WGSL).

> There are several different ways in which you could architect a WebGPU app, but the process will contain the following steps:

Author: Much appreciated. I've taken those suggestions on board, and also made some other changes to make it clear that in a real app, you'll likely be dealing with multiple pipelines.

> There are several different ways in which you could architect a WebGPU app, but the process will contain the following steps:

Reviewer: LGTM

1. [Create shader modules](#create_shader_modules): Write your shader code in WGSL and package it into one or more shader modules.

> 2. [Get and configure the canvas context](#get_and_configure_the_canvas_context): Get the `webgpu` context of a `<canvas>` element and configure it to receive information on what graphics to render from your GPU logical device. This step is not necessary if you are creating a compute pipeline rather than a render pipeline.

Author: Updated; let me know if you'd like to see any other changes.

> 2. [Get and configure the canvas context](#get_and_configure_the_canvas_context): Get the `webgpu` context of a `<canvas>` element and configure it to receive information on what graphics to render from your GPU logical device. This step is not necessary if you are creating a compute pipeline rather than a render pipeline.

Reviewer 2: LGTM

> 3. [Create buffers containing your data](#create_a_buffer_and_write_our_triangle_data_into_it): The data that you want processed by your pipeline needs to be stored in GPU buffers so the app can access it.

Author: Updated

> 4. [Create a pipeline](#define_and_create_the_rendering_pipeline): Define a pipeline descriptor that describes the desired pipeline in detail, including the required data structure, shaders, and resource layout, then create a pipeline from it.

Author: I've handled this one with slightly different wording, in the next commit.

> 4. [Create a pipeline](#define_and_create_the_rendering_pipeline): Define a pipeline descriptor that describes the desired pipeline in detail, including the required data structure, shaders, and resource layout, then create a pipeline from it.

Author: Updated throughout to use pipelines plural

> 4. [Create a pipeline](#define_and_create_the_rendering_pipeline): Define a pipeline descriptor that describes the desired pipeline in detail, including the required data structure, shaders, and resource layout, then create a pipeline from it.

Reviewer 2: LGTM

> 4. [Create a pipeline](#define_and_create_the_rendering_pipeline): Define a pipeline descriptor that describes the desired pipeline in detail, including the required data structure, shaders, and resource layout, then create a pipeline from it.

Reviewer: I like your wording, LGTM!

5. [Run a compute/rendering pass](#running_a_rendering_pass): This involves a number of substeps:
   1. Create a command encoder that can encode a set of commands to be passed to the GPU to execute.
   2. Create a pass encoder object on which commands can be directly issued.
   3. Run commands to specify which pipeline to use, what buffer(s) to get the required data from, how many drawing operations to run (in the case of a render pipeline), etc.
   4. Finalize the command list and encapsulate it in a command buffer.
   5. Submit the command buffer to the GPU via the logical device's command queue.

In the sections below, we will examine a basic render pipeline demo, to allow you to explore what a pipeline requires. Later on, we'll also examine a [basic compute pipeline](#basic_compute_pipeline) example, looking at how it differs from the rendering pipeline.

## Basic render pipeline

In our [basic render demo](https://webgpu-basic-render.glitch.me/) we give a `<canvas>` element a solid blue background and draw a triangle onto it.

### Create shader modules

> We are using the following shader code. The vertex shader (`@vertex` block, which will be the entry point for the vertex pipeline stage) accepts a chunk of data containing a position and a color, positions the vertex according to the given position, then passes the data along to the fragment shader. The fragment shader (`@fragment` block, which will be the entry point for the fragment pipeline stage) accepts the data from the vertex shader and colors the vertex according to the given color.

Reviewer 2: This was to try and point out that each vertex as a different "pure color" (pure red/green/blue) but each fragment will have `fragData.color` be the interpolated value of the `output.color` from vertices. Otherwise it wouldn't be clear what the color of pixels in the middle is.

But this is maybe too much details for this article.

```js
const shaders = `
struct VertexOut {

>     @builtin(position) position : vec4<f32>,

Reviewer 2: nit: the shaders and the text below could use the shorthand `vec4f` instead of `vec4<f32>` (developers don't like typing more than needed)

Reviewer 2: LGTM

    @location(0) color : vec4<f32>
}

@vertex
fn vertex_main(@location(0) position: vec4<f32>,
                 @location(1) color: vec4<f32>) -> VertexOut
{
  var output : VertexOut;
  output.position = position;
  output.color = color;
  return output;
}

@fragment
fn fragment_main(fragData: VertexOut) -> @location(0) vec4<f32>
{
  return fragData.color;
}
`;
```

> > **Note:** In our demos we are storing our shader code inside a template literal, but you can store it anywhere from which it can easily be retrieved as text to be fed into your WebGPU program. For example, another common practice is to store shaders inside a {{htmlelement("script")}} element with a `type` attribute set to a custom value to designate it as a data block, and then retrieve the contents using {{domxref("Node.textContent")}}.

Author: Great stuff! I've updated the note to mention this.

> > **Note:** In our demos we are storing our shader code inside a template literal, but you can store it anywhere from which it can easily be retrieved as text to be fed into your WebGPU program. For example, another common practice is to store shaders inside a {{htmlelement("script")}} element with a `type` attribute set to a custom value to designate it as a data block, and then retrieve the contents using {{domxref("Node.textContent")}}.

Reviewer 2: LGTM

To make your shader code available to WebGPU, you have to put it inside a {{domxref("GPUShaderModule")}} via a {{domxref("GPUDevice.createShaderModule()")}} call, passing your shader code as an property inside a descriptor object. For example:

```js
const shaderModule = device.createShaderModule({
  code: shaders
});
```

### Get and configure the canvas context

In a render pipeline, we need to specify somewhere to render the graphics to. In this case we are getting a reference to an onscreen `<canvas>` element then calling {{domxref("HTMLCanvasElement.getContext()")}} with a parameter of `webgpu` to return its GPU context (a {{domxref("GPUCanvasContext")}} instance).

> From there, we configure the context's swap chain, which is the source of the graphical information to be rendered. This is done with a call to {{domxref("GPUCanvasContext.configure()")}}, passing it an options object containing the {{domxref("GPUDevice")}} that the rendering information will come from, the format the textures will have, and the alpha mode to use when rendering semi-transparent textures.

Author: Ah, cool. This did seem to be introducing excess terminology without much benefit. I've removed it and simplified the section.

> From there, we configure the context's swap chain, which is the source of the graphical information to be rendered. This is done with a call to {{domxref("GPUCanvasContext.configure()")}}, passing it an options object containing the {{domxref("GPUDevice")}} that the rendering information will come from, the format the textures will have, and the alpha mode to use when rendering semi-transparent textures.

Reviewer 2: LGTM

```js
const canvas = document.querySelector('#gpuCanvas');
const context = canvas.getContext('webgpu');

// Configure swap chain, link the GPU to the canvas
context.configure({
    device: device,

>     format: 'bgra8unorm',

Reviewer: To demonstrate best practices, please use `navigator.gpu.getPreferredCanvasFormat()` here. It will select the most efficient format (either 'bgra8unorm' or 'rgba8unorm') for the user's device. Statically selecting a canvas format like this will likely incur an extra texture copy on mobile devices.

Author: OK, I've updated this in the article text (in next commit) and in the demo, and also added a note about it being best practice.

Reviewer: LGTM

>     format: 'bgra8unorm',

Author: Yup, @toji commented on this too. I've updated the text and the demo.

>     format: 'bgra8unorm',

Reviewer 2: LGTM

```js
const vertices = new Float32Array([
    0.0,  0.6, 0, 1, 1, 0, 0, 1,
   -0.5, -0.6, 0, 1, 0, 1, 0, 1,
    0.5, -0.6, 0, 1, 0, 0, 1, 1
]);
```

However, we've got an issue here. We need to get our data into a {{domxref("GPUBuffer")}}. Behind the scenes, this type of buffer is stored in memory very tightly integrated with the GPU's cores to allow for the high performance required for processing. As a side effect, this memory can't be accessed by processes running on the host system, like the browser.

The {{domxref("GPUBuffer")}} is created via a call to {{domxref("GPUDevice.createBuffer()")}}. We give it a size equal to the length of the `vertices` array so it can contain all the data, `VERTEX` and `COPY_DST` flags to indicate that the buffer will be used as a vertex buffer and the destination of copy operations, and we also specify that it should be mapped at creation.

```js
const vertexBuffer = device.createBuffer({
    size: vertices.byteLength, // make it big enough to store vertices in

>     usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,

Author: I've updated this to use `writeBuffer()`, so I'm keeping it in ;-)

>     usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,

Reviewer 2: LGTM

    mappedAtCreation: true // mapped at creation
});
```

> The third option parameter, `mappedAtCreation: true`, is key here — this causes the buffer to be immediately mapped to an intermediate memory block that is not as performant, but can be accessed by the host system to be written to. If you don't map the buffer at creation, you can still map it later using {{domxref("GPUBuffer.mapAsync()")}}.

Author: I think we are OK here; this section no longer uses `mappedAtCreation`. Let me know if you think the update looks Ok.

> The third option parameter, `mappedAtCreation: true`, is key here — this causes the buffer to be immediately mapped to an intermediate memory block that is not as performant, but can be accessed by the host system to be written to. If you don't map the buffer at creation, you can still map it later using {{domxref("GPUBuffer.mapAsync()")}}.

Reviewer 2: LGTM

With the buffer unlocked so it can be modified, we return a range of data that has been mapped using {{domxref("GPUBuffer.getMappedRange()")}} (which returns all the mapped data by default), put it into another {{jsxref("Float32Array")}}, then set its contents to the contents of the `vertices` array, using {{domxref("TypedArray.set()")}}.

```js
new Float32Array(vertexBuffer.getMappedRange()).set(vertices);
```

Now we have finished modifying the buffer contents, we need to unmap the buffer, which copies the updated data back to the GPU. This is done using {{domxref("GPUBuffer.unmap()")}}

```js
vertexBuffer.unmap();

> ```

Author: Ha, I missed this convenience method! It took me a couple of minutes to rewrite my render demo to use this, and I've updated the article too, mentioning that we also cover mapping later in the compute example. 

I like how this has worked out — we show them the most straightforward path first (which is a lot more logical than all the mapping stuff), then mention that mapping is also available and demonstrate it later on.

```js
const vertexBuffers = [{
    attributes: [{
        shaderLocation: 0, // position
        offset: 0,
        format: 'float32x4'
    }, {
        shaderLocation: 1, // color
        offset: 16,
        format: 'float32x4'
    }],
    arrayStride: 32,
    stepMode: 'vertex'
}];
```

Next, we create a descriptor object that specifies the configuration of our render pipeline stages. For both the shader types, we specify that the {{domxref("GPUShaderModule")}} the required shaders can be found in is `shaderModule`, and the name of the function that acts as the entry point for each shader.

In addition, in the case of the vertex shader we provide our `vertexBuffers` object to provide the expected state of our vertex data. And in the case of our fragment shader, we provide an array of color target states that indicate the specified rendering format (this matches the format specified in our canvas context config earlier).

> We also specify a `primitive` state, which in this case just states the type of primitive we will be drawing, and a `layout` of `auto`. The `layout` property defines the layout (structure, purpose, and type) of all the GPU resources (buffers, textures, etc.) used during the execution of the pipeline. In more complex apps, this would take the form of a {{domxref("GPUBindGroupLayout")}} object, created using {{domxref("GPUDevice.createBindGroupLayout()")}} (you can see an example in our [Basic compute pipeline](#basic_compute_pipeline)), which allows the GPU to figure out how to run the pipeline most efficiently ahead of time. Here however we are specifying the `auto` value, which will cause the pipeline to use a default layout.

Reviewer 2: nit: the layout would be a GPUPipelineLayout made of GPUBindGroupLayouts. Making explicit layouts can have two benefits: make the application less brittle to the shader statically using bindings or not (if you stop referencing the binding it will no longer appear in the layout and will cause a validation error if you specify it in the bindgroup) or to allow better reuse of bindgroups between pipelines (so that a bind group for scene-global data can be created once at the start of the frame instead of once per object for example).

Reviewer 2: nit: the term is "implicit layout" instead of "default layout".

Author: Ah, I see what I've done here; I just got muddled a bit when explaining the code structure. I've edited that part, and also made sure to mention "implicit".

Reviewer 2: Thanks! LGTM

> We also specify a `primitive` state, which in this case just states the type of primitive we will be drawing, and a `layout` of `auto`. The `layout` property defines the layout (structure, purpose, and type) of all the GPU resources (buffers, textures, etc.) used during the execution of the pipeline. In more complex apps, this would take the form of a {{domxref("GPUBindGroupLayout")}} object, created using {{domxref("GPUDevice.createBindGroupLayout()")}} (you can see an example in our [Basic compute pipeline](#basic_compute_pipeline)), which allows the GPU to figure out how to run the pipeline most efficiently ahead of time. Here however we are specifying the `auto` value, which will cause the pipeline to use a default layout.

Reviewer: ```suggestion
We also specify a `primitive` state, which in this case just states the type of primitive we will be drawing, and a `layout` of `auto`. The `layout` property defines the layout (structure, purpose, and type) of all the GPU resources (buffers, textures, etc.) used during the execution of the pipeline. In more complex apps, this would take the form of a {{domxref("GPUBindGroupLayout")}} object, created using {{domxref("GPUDevice.createBindGroupLayout()")}} (you can see an example in our [Basic compute pipeline](#basic_compute_pipeline)), which allows the GPU to figure out how to run the pipeline most efficiently ahead of time. Here however we are specifying the `auto` value, which will cause the pipeline to generate a bind group layout for us based on any bindings defined in the shader code.
```

```js
const pipelineDescriptor = {
    vertex: {
        module: shaderModule,
        entryPoint: 'vertex_main',
        buffers: vertexBuffers
    },
    fragment: {
        module: shaderModule,
        entryPoint: 'fragment_main',
        targets: [{

>         format: 'bgra8unorm'

Reviewer 2: nit: missing indentation?

Author: Corrected; thanks!

Reviewer 2: LGTM

>         format: 'bgra8unorm'

Reviewer: Same request as earlier: It's good that the text points out that this matches the canvas context config, but please make sure it's also pulling it's value from `navigator.gpu.getPreferredCanvasFormat()`.

Author: See above comment

        }]
    },
    primitive: {
        topology: 'triangle-list'
    },
    layout: 'auto'
};
```

Finally, we can create a {{domxref("GPURenderPipeline")}} based on our `pipelineDescriptor` object, by passing it in as a parameter to a {{domxref("GPUDevice.createRenderPipeline()")}} method call.

```js
const renderPipeline = device.createRenderPipeline(pipelineDescriptor);
```

### Running a rendering pass

Now that all the setup is done, we can actually run a rendering pass and draw something onto our `<canvas>`. To encode any commands to be later issued to the GPU, you need to create a {{domxref("GPUCommandEncoder")}} instance, which is done using a {{domxref("GPUDevice.createCommandEncoder()")}} call.

```js
const commandEncoder = device.createCommandEncoder();
```

Next up we start the rendering pass running by creating a {{domxref("GPURenderPassEncoder")}} instance with a {{domxref("GPUCommandEncoder.beginRenderPass()")}} call. This method takes a descriptor object as a parameter, the only mandatory property of which is a `colorAttachments` array. In this case, we specify:

> 1. A view to render into the `<canvas>`; we create a new view via {{domxref("GPUTexture.createView", "context.getCurrentTexture().createView()")}}.

Reviewer: ```suggestion
1. A texture view to render into; we create a new view from the `<canvas>` via {{domxref("GPUTexture.createView", "context.getCurrentTexture().createView()")}}.
```

2. That the view should be "cleared" to a specified color once loaded and before any drawing takes place. This is what causes the blue background behind the triangle.
3. That the value of the current rendering pass should be stored for this color attachment.

```js
const clearColor = { r: 0.0, g: 0.5, b: 1.0, a: 1.0 };

const renderPassDescriptor = {
    colorAttachments: [{
        clearValue: clearColor,

>         loadOp: 'clear',

Reviewer 2: Thinking of it, better names would have been "startOp" and "endOp" but I think all APIs call them load and store... Oh well it's probably too late to change and not important enough :)

Author: ¯\_(ツ)_/¯

It is far from the worst naming choice I've seen in an API ;-)

        storeOp: 'store',
        view: context.getCurrentTexture().createView()
    }]
};

const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
```

Now we can invoke methods of the rendering pass encoder to draw our triangle:

1. {{domxref("GPURenderPassEncoder.setPipeline()")}} is called with our `renderPipeline` object as a parameter to specify the pipeline to use for the rendering pass.

> 2. {{domxref("GPURenderPassEncoder.setVertexBuffer()")}} is called with our `vertexBuffer` object as a parameter to act as the data source to pass to the pipeline to render. The first parameter is the slot to set the vertex buffer for, and is a reference to the equivalent `shaderLocation` value in the `vertexBuffers` object.

Reviewer: ```suggestion
2. {{domxref("GPURenderPassEncoder.setVertexBuffer()")}} is called with our `vertexBuffer` object as a parameter to act as the data source to pass to the pipeline to render. The first parameter is the slot to set the vertex buffer for, and is a reference to the index of the element in the `vertexBuffers` array which describes this buffer's layout.
```

1. We invoke the {{domxref("GPURenderPassEncoder.end()")}} method to signal the end of the render pass command list.
2. We invoke the {{domxref("GPUCommandEncoder.finish()")}} method to complete recording of the issued command sequence and encapsulate it into a {{domxref("GPUCommandBuffer")}} object instance.
3. We submit the {{domxref("GPUCommandBuffer")}} to the device's command queue (represented by a {{domxref("GPUQueue")}} instance) to be sent to the GPU. The device's queue is available via the {{domxref("GPUDevice.queue")}} property, and an array of {{domxref("GPUCommandBuffer")}} instances can be added to the queue via a {{domxref("GPUQueue.submit()")}} call.

These three steps can be achieved via the following two lines:

```js
passEncoder.end();

device.queue.submit([commandEncoder.finish()]);
```

## Basic compute pipeline

In our [basic compute demo](https://webgpu-basic-compute.glitch.me/), we get the GPU to calculate some values, store them in an output buffer, copy the data across to a staging buffer, then map that staging buffer so that the data can be read out to JavaScript and logged to the console.

> The app follows a similar structure to the basic rendering demo. We create a {{domxref("GPUDevice")}} reference in the same way as before, and encapsulate our shader code into a {{domxref("GPUShaderModule")}} via a {{domxref("GPUDevice.createShaderModule()")}} call. The difference here is that our shader code only has one shader, a compute shader:

Reviewer: ```suggestion
The app follows a similar structure to the basic rendering demo. We create a {{domxref("GPUDevice")}} reference in the same way as before, and encapsulate our shader code into a {{domxref("GPUShaderModule")}} via a {{domxref("GPUDevice.createShaderModule()")}} call. The difference here is that our shader code only has one shader stage, a compute shader:
```

Nit: Technically the other example also only had one shader, but the shader had two stages (vertex and fragment)

Author: I've had a look through and made some updates to the text to reflect this.

```js
const shader = `
@group(0) @binding(1)
var<storage, read_write> output: array<f32>;

@compute @workgroup_size(64)
fn main(

  @builtin(global_invocation_id)

>   global_id : vec3<u32>,

Reviewer 2: nit: vec3u

Author: updated.

  @builtin(local_invocation_id)
  local_id : vec3<u32>,

) {

>   output[global_id.x] =

Reviewer 2: Technically this shader access the array out of bounds at indices 1001..1023 in the last workgroup of the dispatch. WebGPU makes this safe (by clamping array indices to 1000 in this case) but maybe to help developers think of this, there should be an early out in the shader for this case?

Author: I'm not really sure how to do that; can you write a code update for me? Or is there a way to avoid this issue, e.g. by changing the workgroup size?

Reviewer 2: Made a small suggestion. It might be too head-scratching though.

    f32(global_id.x) * 1000. + f32(local_id.x);
}
`;
```

### Create buffers to handle our data

In this example we create two {{domxref("GPUBuffer")}} instances to handle our data, an `output` buffer to write the GPU calculation results directly to at high speed, and a `stagingBuffer` that we'll copy the `output` contents to, which can be mapped to allow JavaScript to access the values.

- `output` is specified as a storage buffer that will be the source of a copy operation.
- `stagingBuffer` is specified as a buffer that can be mapped for reading by JavaScript, and will be the destination of a copy operation.

```js
const BUFFER_SIZE = 1000;

const output = device.createBuffer({
    size: BUFFER_SIZE,
    usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
});

const stagingBuffer = device.createBuffer({
    size: BUFFER_SIZE,
    usage: GPUBufferUsage.MAP_READ | GPUBufferUsage.COPY_DST,
});
```

### Create a bind group layout

> When the pipeline is created, instead of using a `layout` of `auto`, we specify a bind group to use for the pipeline. This involves first creating a {{domxref("GPUBindGroupLayout")}} (via a call to {{domxref("GPUDevice.createBindGroupLayout()")}}) that defines the structure and purpose of GPU resources such as buffers that will be used in this pipeline. This layout is used as a template for bind groups to adhere to. In this case we give the pipeline access to a single memory buffer, tied to binding slot 1 (this matches the relevant binding number in our shader code — `@binding(1)`), usable in the compute stage of the pipeline, and with the buffer's purpose defined as `storage`.

Reviewer 2: nit: Technically an implicit layout can be used even if a bind group is needed: the layout can be retrieved with `pipeline.getBindGroupLayout`. Maybe this could explain that we do it explicitly to show how it works?

Author: I've added a note about it, good call.

```js
const bindGroupLayout =
device.createBindGroupLayout({
    entries: [{

>         binding: 1,

Reviewer 2: nit: why not binding 0? Indices for bindings are in the [0, SomeLargeValue) range.

Author: I have no good answer for that ;-)

I've changed it to 0, in the article and demo code.

Reviewer 2: LGTM

        visibility: GPUShaderStage.COMPUTE,
        buffer: {
            type: "storage",
        }
    }]
});
```

> Next we create a {{domxref("GPUBindGroup")}} by calling {{domxref("GPUDevice.createBindGroup()")}}. We passing this method call a descriptor object that specifies the bind group layout to base this bind group on, and the details of the variable to bind to the slot defined in the layout. In this case, we are declaring bind group 1, and specifying that it should be written to the `output` buffer we defined earlier).

Reviewer: ```suggestion
Next we create a {{domxref("GPUBindGroup")}} by calling {{domxref("GPUDevice.createBindGroup()")}}. We pass this method call a descriptor object that specifies the bind group layout to base this bind group on, and the details of the variable to bind to the slot defined in the layout. In this case, we are declaring binding 1, and specifying that it should be written to the `output` buffer we defined earlier).
```

```js
const bindGroup = device.createBindGroup({
    layout: bindGroupLayout,
    entries: [{
        binding: 1,
        resource: {

>         buffer: output,

Reviewer 2: nit: indentation

Author: thanks; updated.

        },
    }],
});
```

### Create a compute pipeline

With the above all in place, we can now create a compute pipeline by calling {{domxref("GPUDevice.createComputePipeline()")}}, passing it a pipeline descriptor object. This works in a similar way to creating a render pipeline. We describe the compute shader, specifying what module to find the code in and what the entry point is. We also specify a `layout` for the pipeline, in this case creating a layout based on the `bindGroupLayout` we defined earlier via a {{domxref("GPUDevice.createPipelineLayout()")}} call.

```js
const computePipeline = device.createComputePipeline({
    layout: device.createPipelineLayout({
        bindGroupLayouts: [bindGroupLayout]
    }),
    compute: {
      module: shaderModule,
      entryPoint: 'main'
    },
});
```

> One difference here from the pipeline layout is that we are not specifying a primitive type, as we are not drawing anything.

Reviewer: ```suggestion
One difference here from the render pipeline layout is that we are not specifying a primitive type, as we are not drawing anything.
```

```js
// Copy output buffer to staging buffer
commandEncoder.copyBufferToBuffer(
    output,
    0, // Source offset
    stagingBuffer,
    0, // Destination offset
    BUFFER_SIZE
);

// End frame by passing array of command buffers to command queue for execution
device.queue.submit([commandEncoder.finish()]);
```

Once the output data is available in the `stagingBuffer`, we use the {{domxref("GPUBuffer.mapAsync()")}} method to map the data to intermediate memory, grab a reference to the mapped range using {{domxref("GPUBuffer.getMappedRange()")}}, copy the data into JavaScript, and then log it to the console. We also unmap the `stagingBuffer` once we are finished with it.

```js
// map staging buffer to read results back to JS
await stagingBuffer.mapAsync(
    GPUMapMode.READ,
    0, // Offset
    BUFFER_SIZE // Length
);

const copyArrayBuffer = stagingBuffer.getMappedRange(0, BUFFER_SIZE);
const data = copyArrayBuffer.slice();
stagingBuffer.unmap();
console.log(new Float32Array(data));
```

## GPU error handling

> WebGPU calls are validated asynchronously on the GPU. If errors are found, the problem call is marked as invalid on the GPU side. If another call is made that relies on the return value of an invalidated call, that object will also be marked as invalid, and so on. For this reason, errors in WebGPU are referred to as "contagious".

Reviewer 2: nit: not exactly on the GPU but in the GPU process (I think the WebGPU explainer has more info on this).

Author: I've updated the first sentence to "WebGPU calls are validated asynchronously in the GPU process." in my next commit. I link to the explainer a couple of times at the end of the section for more info. I wanted to be fairly brief here, as the article is long enough already.

  - : Based on a {{domxref("GPUBindGroupLayout")}}, a `GPUBindGroup` defines a set of resources to be bound together in a group and how those resources are used in shader stages.
- {{domxref("GPUBindGroupLayout")}}
  - : Defines the structure and purpose of related GPU resources such as buffers that will be used in a pipeline, and is used as a template when creating {{domxref("GPUBindGroup")}}s.
- {{domxref("GPUComputePipeline")}}
  - : Controls the compute shader stage and can be used in a {{domxref("GPUComputePassEncoder")}}.
- {{domxref("GPUPipelineLayout")}}
  - : Defines the {{domxref("GPUBindGroupLayout")}}s used by a pipeline. {{domxref("GPUBindGroup")}}s used with the pipeline during command encoding must have compatible {{domxref("GPUBindGroupLayout")}}s.

### Encoding and submitting commands to the GPU

- {{domxref("GPUCommandBuffer")}}
  - : Represents a pre-recorded list of GPU commands that can be submitted to a {{domxref("GPUQueue")}} for execution. Created by calling {{domxref("GPUCommandEncoder.finish()")}}.
- {{domxref("GPUCommandEncoder")}}
  - : Represents a command encoder, used to encode commands to be issued to the GPU.
- {{domxref("GPUComputePassEncoder")}}

>   - : Encodes commands issued related to controlling the compute shader stage, as issued by a {{domxref("GPUComputePipeline")}}. Part of the overall encoding activity of a {{domxref("GPUCommandEncoder")}}.

Reviewer: This "issued" felt out of place

```suggestion
  - : Encodes commands related to controlling the compute shader stage, as issued by a {{domxref("GPUComputePipeline")}}. Part of the overall encoding activity of a {{domxref("GPUCommandEncoder")}}.
```

Author: Yeah, that does look a bit weird; I've removed it in my local copy

- {{domxref("GPUQueue")}}
  - : controls execution of encoded commends on the GPU.
- {{domxref("GPURenderBundle")}}
  - : A container for pre-recorded bundles of commands (see {{domxref("GPURenderBundleEncoder")}}).
- {{domxref("GPURenderBundleEncoder")}}
  - : Used to pre-record bundles of commands. These can be reused in {{domxref("GPURenderPassEncoder")}}s via the {{domxref("GPURenderPassEncoder.executeBundles", "executeBundles()")}} method, as many times as required.
- {{domxref("GPURenderPassEncoder")}}

>   - : Encodes commands issued related to controlling the vertex and fragment shader stages, as issued by a {{domxref("GPURenderPipeline")}}. Part of the overall encoding activity of a {{domxref("GPUCommandEncoder")}}.

Reviewer: Ditto

```suggestion
  - : Encodes commands related to controlling the vertex and fragment shader stages, as issued by a {{domxref("GPURenderPipeline")}}. Part of the overall encoding activity of a {{domxref("GPUCommandEncoder")}}.
```

Author: removed

- {{domxref("GPUInternalError")}}
  - : One of the types of errors surfaced by {{domxref("GPUDevice.popErrorScope")}} and the {{domxref("GPUDevice.uncapturederror_event", "uncapturederror")}} event. Indicates than an operation failed for a system or implementation-specific reason, even when all validation requirements were satisfied.
- {{domxref("GPUOutOfMemoryError")}}
  - : One of the types of errors surfaced by {{domxref("GPUDevice.popErrorScope")}} and the {{domxref("GPUDevice.uncapturederror_event", "uncapturederror")}} event. Indicates that there was not enough free memory to complete the requested operation.
- {{domxref("GPUPipelineError")}}
  - : Describes a pipeline failure. The value received when a {{jsxref("Promise")}} returned by a {{domxref("GPUDevice.createComputePipelineAsync()")}} or {{domxref("GPUDevice.createRenderPipelineAsync()")}} call rejects.
- {{domxref("GPUUncapturedErrorEvent")}}
  - : The event object type for the {{domxref("GPUDevice")}} {{domxref("GPUDevice.uncapturederror_event", "uncapturederror")}} event.
- {{domxref("GPUValidationError")}}

>   - : One of the types of errors surfaced by {{domxref("GPUDevice.popErrorScope")}} and the {{domxref("GPUDevice.uncapturederror_event", "uncapturederror")}} event. Describes an application error indicating that an operation did not pass the app's validation constraints.

Reviewer: Minor clarification

```suggestion
  - : One of the types of errors surfaced by {{domxref("GPUDevice.popErrorScope")}} and the {{domxref("GPUDevice.uncapturederror_event", "uncapturederror")}} event. Describes an application error indicating that an operation did not pass the WebGPU API's validation constraints.
```

Author: Updated in my local copy

## Examples

- [Basic compute demo](https://webgpu-basic-compute.glitch.me/)
- [Basic render demo](https://webgpu-basic-render.glitch.me/)

> - [WebGPU samples](https://austin-eng.com/webgpu-samples/)

Reviewer: These have recently moved!
```suggestion
- [WebGPU samples](https://webgpu.github.io/webgpu-samples/)
```

Author: Updated!

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [WebGPU best practices](https://toji.github.io/webgpu-best-practices/)

Reviewer: So have these!
```suggestion
- [WebGPU best practices](https://toji.dev/webgpu-best-practices/)
```

Author: updated these too. Thanks for having a look.

- {{domxref("GPUQueue")}}
  - : controls execution of encoded commands on the GPU.
- {{domxref("GPURenderBundle")}}
  - : A container for pre-recorded bundles of commands (see {{domxref("GPURenderBundleEncoder")}}).
- {{domxref("GPURenderBundleEncoder")}}
  - : Used to pre-record bundles of commands. These can be reused in {{domxref("GPURenderPassEncoder")}}s via the {{domxref("GPURenderPassEncoder.executeBundles", "executeBundles()")}} method, as many times as required.
- {{domxref("GPURenderPassEncoder")}}
  - : Encodes commands related to controlling the vertex and fragment shader stages, as issued by a {{domxref("GPURenderPipeline")}}. Part of the overall encoding activity of a {{domxref("GPUCommandEncoder")}}.
- {{domxref("GPURenderPipeline")}}
  - : Controls the vertex and fragment shader stages and can be used in a {{domxref("GPURenderPassEncoder")}} or {{domxref("GPURenderBundleEncoder")}}.

### Running queries on rendering passes

- {{domxref("GPUQuerySet")}}
  - : Used to record the results of queries on passes, such as occlusion or timestamp queries.

### Debugging errors

- {{domxref("GPUCompilationInfo")}}

>   - : An array of {{domxref("GPUCompilationMessage")}} objects, generated by the GPU shader module compiler to help diagnose problems with shader code. Accessed via {{domxref("GPUShaderModule.compilationInfo()")}}.

Reviewer: This changed in https://github.com/gpuweb/gpuweb/pull/3857 (And it is my sincere hope that that's the last of the name changes we'll get prior to shipping!)

Implementation may not have landed in Chrome just yet.

```suggestion
  - : An array of {{domxref("GPUCompilationMessage")}} objects, generated by the GPU shader module compiler to help diagnose problems with shader code. Accessed via {{domxref("GPUShaderModule.getCompilationInfo()")}}.
```

Author: Ah, no worries. I've updated it, and will bear it in mind when testing as I write the doc for it.
